core/chain/src/lib.rs

```diff
@@ -1,27 +1,64 @@
 extern crate parking_lot;
 extern crate primitives;
+extern crate serde;
 extern crate storage;
 
 use std::collections::HashMap;
+use std::fmt::Debug;
 use std::sync::Arc;
 
 use parking_lot::RwLock;
+use serde::{de::DeserializeOwned, Serialize};
 
 use primitives::hash::CryptoHash;
-use primitives::traits::{Block, Decode, Encode, Header};
-use primitives::types::BlockId;
+use primitives::traits::{Decode, Encode, Signer};
+use primitives::types::{BlockId, PartialSignature};
 use primitives::utils::index_to_bytes;
 use storage::Storage;
 
-const BLOCKCHAIN_GENESIS_BLOCK: &[u8] = b"genesis";
 const BLOCKCHAIN_BEST_BLOCK: &[u8] = b"best";
 
+/// Trait that abstracts ``Header"
+pub trait Header: Debug + Clone + Send + Sync + Serialize + DeserializeOwned + Eq + 'static
+{
+    /// Returns hash of the block body.
+    fn block_hash(&self) -> CryptoHash;
+
+    /// Returns block index.
+    fn index(&self) -> u64;
+
+    /// Returns hash of parent block.
+    fn parent_hash(&self) -> CryptoHash;
+}
+
+/// Trait that abstracts a ``Block", Is used for both beacon-chain blocks
+/// and shard-chain blocks.
+pub trait Block: Debug + Clone + Send + Sync + Serialize + DeserializeOwned + Eq + 'static {
+    type Header: Header;
+
+    /// Returns signed header for given block.
+    fn header(&self) -> Self::Header;
+
+    /// Returns hash of the block body.
+    fn block_hash(&self) -> CryptoHash;
+
+    /// Signs this block with given signer and returns part of multi signature.
+    fn sign(&self, signer: Arc<Signer>) -> PartialSignature {
+        signer.sign(&self.block_hash())
+    }
+
+    /// Add signature to multi sign.
+    fn add_signature(&mut self, signature: PartialSignature);
+}
+
 /// General BlockChain container.
 pub struct BlockChain<B: Block> {
     /// Storage backend.
     storage: Arc<Storage>,
     /// Genesis hash
     pub genesis_hash: CryptoHash,
+    /// Best block key of current blockchain.
+    best_block_key: Vec<u8>,
```

Comment:
If we know the exact size of the key then we should be using a fixed-size array of `u8`.

Reply:
I added size, even though I personally find that weird. 36 there makes no sense.

Reply:
also now need:
```
        let mut best_block_key = [0; 36];
        let vec_key: Vec<u8> = genesis_hash.as_ref().iter().chain(BLOCKCHAIN_BEST_BLOCK).cloned().collect();
        best_block_key.copy_from_slice(vec_key.as_ref());
```

Reply:
or
```
        let mut best_block_key = [0; 36];
        best_block_key[..32].copy_from_slice(genesis_hash.as_ref());
        best_block_key[33..36].copy_from_slice(BLOCKCHAIN_BEST_BLOCK);
```